lib/3scale/backend/job_fetcher.rb

```diff
@@ -11,49 +11,41 @@ class JobFetcher
       REDIS_TIMEOUT = 60
       private_constant :REDIS_TIMEOUT
 
-      RedisConnectionError = Class.new(RuntimeError)
-
       # The default redis_client is the one defined in Resque::Helpers
       def initialize(redis_client: redis, fetch_timeout: REDIS_TIMEOUT)
         @redis = redis_client
         @fetch_timeout = fetch_timeout
         @queues ||= QUEUES.map { |q| "queue:#{q}" }
       end
 
-      def fetch
-        encoded_job = pop_from_queue
-        return nil if encoded_job.nil? || encoded_job.empty?
-
-        begin
-          # Resque::Job.new accepts a queue name as a param. It is very
-          # important to set here the same name as the one we set when calling
-          # Resque.enqueue. Resque.enqueue uses the @queue ivar in
-          # BackgroundJob classes as the name of the queue, and then, it stores
-          # the job in a queue called resque:queue:_@queue_. 'resque:' is the
-          # namespace and 'queue:' is added automatically. That's why we need
-          # to call blpop on 'queue:#{q}' above. However, when creating the job
-          # we cannot set 'queue:#{q}' as the name. Otherwise, if it fails and
-          # it is re-queued, it will end up in resque:queue:queue:_@queue_
-          # instead of resque:queue:_@queue_.
-          encoded_job[0].sub!('queue:', '')
-          Resque::Job.new(encoded_job[0],
-                          Yajl::Parser.parse(encoded_job[1], check_utf8: false))
-        rescue Exception => e
-          # I think that the only exception that can be raised here is
-          # Yajl::ParseError. However, this is a critical part of the code so
-          # we will capture all of them just to be safe.
-          Worker.logger.notify(e)
-          nil
-        end
+      # @param wait [Boolean] wait for a job to appear in a queue instead of returning immediately
+      # @param max [Integer] maximum number of records to return
+      #   note: presently ignored in waiting mode because BLMPOP was added in Redis 7
+      #   note: also we need to update redis-rb gem so that sync client can support this parameter
+      # @return [nil, Resque::Job, Array<Resque::Job>] nil when no jobs are present in the queue,
+      #                                                 a single job when max was not specified,
+      #                                                 an array of jobs otherwise
+      def fetch(wait: true, max: nil)
+        encoded_jobs = wait ? wait_pop_from_queue : try_pop_from_queue(max)
+        return if !encoded_jobs || encoded_jobs.empty?
+
+        jobs = encoded_jobs.map { decode_job _1 }
```

Comment:
This code is wrong,because `encoded_jobs` will be a single dimension array like 

`['queue:priority', '{"class": "ReporJob", ... }']` 

when only one job is returned, but a two-dimensional array when there are more jobs:

`[['queue:priority', '{"class": "ReporJob", ... }'],['queue:priority', '{"class": "ReporJob", ... }']]`. 

So `_1` here will be `queue:priority` in the first case and `['queue:priority', '{"class": "ReporJob", ... }']` in the second.

Note how the code to decode the job was like this before:

```ruby
encoded_job[0].sub!('queue:', '')
Resque::Job.new(encoded_job[0],
                          Yajl::Parser.parse(encoded_job[1], check_utf8: false))
```
The calls to `encoded_job[0]` and `encoded_job[1]` prove the decoder expected an array (the second form), but the new decoder expects aparently a `String`, which won't work.

Reply:
I've got it already, yes